### PR TITLE
feat(pairing): add pairing docs

### PR DESCRIPTION
**Pull request by vladikoff**
@shane-tomlinson @rfk r?

### CODE REVIEW

File: docs/pairing-architecture.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+* We use explicit confirmation screens to reduce the risk of phishing or over-the-shoulder attacks.
+
+The Mobile device sends what is essentially an OAuth request, delivered over the websocket channel rather than opened as a local webpage. 
+It includes information such as:
+```
+client_id - its registered OAuth client id
+state - OAuth state parameter
+scope - requested OAuth scopes
+code_challenge and code_challenge_method - OAuth PKCE parameters
```
Comment (shane-tomlinson) [inline docs/pairing-architecture.md]: Can you put a blurb about why the PKCE and keys_jwk are passed in.

File: docs/pairing-architecture.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+* There's no set confirmation order: the Supp and Auth will show the confirmations screens at the same time, and can be accepted in any order.
+
+For additional security, the Channel Server annotates each message with information about the originating device, such as geoip lookup data. This may help the user detect if they're connecting to a different device than expected.
+
+#### 4) Desktop device authorizes OAuth request
+
+When the user accepts the confirmation screen on the Desktop device, it will use its local credentials to authorize the OAuth request from the Mobile device. This involves no user interaction, and will include the following steps:
+
+* Check that redirect_uri is the expected value; error out if not.
+* Check that scope contains exactly `"profile"` and `"https://identity.mozilla.com/apps/oldsync"`; error out if not.
```
Comment (shane-tomlinson) [inline docs/pairing-architecture.md]: Have these checks been done?

File: docs/pairing-architecture.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
+## Pairing Flow Architecture
+
+To allow users to quickly connect a mobile browser to Firefox Sync, without having to type their password,
+by pairing with an authenticated Firefox profile on their PC via scanning a QR code.
+See the [Project Plan Document](https://docs.google.com/document/d/1mFf0sfEK8o1csXLeyK1x4GS9SUMIq4q8bu25LiydPew/edit#) for more information.
+
+The Desktop and Mobile devices will not communicate directly, because we cannot assume that they will be on
+the same network and because this is difficult to achieve from web content.
+Instead, they will exchange encrypted messages via websocket to a lightweight "channel server" service which can relay messages back and forth:
+
```
Comment (shane-tomlinson) [inline docs/pairing-architecture.md]: This intro could be expanded upon to include info about what information the new device needs and how that info is received from the client.

In essence we are doing a remote authorization where the already authenticated device generates Sync keys and an OAuth code for the new device. PKCE facilitates the transfer of the Sync key between devices and the channel server facilitates direct communication between the two, including the final transfer of the OAuth code.

The security properties section from the feature doc would also be useful.

Comment (rfk) [reply docs/pairing-architecture.md]: "remote authorization" is a great way to describe this!

File: docs/pairing-architecture.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
+## Pairing Flow Architecture
+
+To allow users to quickly connect a mobile browser to Firefox Sync, without having to type their password,
+by pairing with an authenticated Firefox profile on their PC via scanning a QR code.
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: 👍

File: docs/pairing-architecture.md (hunk 3 above)
Comment (vladikoff) [reply docs/pairing-architecture.md]: 👍

File: docs/pairing-architecture.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+It opens a websocket to the Channel Server using the same fxa-pairing-channel library specifying the channel id, channel key.
+The Channel Server will now relay websocket messages from Mobile to Desktop and vice-versa. The supplicant broker handles the
+creation of the PairingChannelClient and the state machine. The supplicant triggers messages after user's approval and finally sends
+the code to the relier by using the existing OAuth redirection broker.
+
+#### 3) Devices exchange metadata and show confirmation screens
+
+At this point the devices have established a baseline level of trust and can exchange metadata to allow the user to confirm the pairing on both sides. 
+
+* We use explicit confirmation screens to reduce the risk of phishing or over-the-shoulder attacks.
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: sometimes I make single point bulleted lists, it's bad habit

File: docs/pairing-architecture.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+The Firefox Desktop native code gathers information from web content via WebChannels and communicates that to the supplicant using the 
+channel server socket.
+
+Meanwhile, the web content opened on the supplicant device connects directly to a WebSocket, without native code infrastructure.
+It waits for user's approval on both sides of the flow and once that is achieved, it can receive the OAuth `code` and `state` 
+from the authority side of the pairing flow.
+
+### Pairing Channel messages
+The message payload is a JSON object that contains a message identifier and a data object.
+Every message envelope also contains information about the sender (IP address, geolocation..).
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: Yea! Adding

File: docs/pairing-architecture.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+The Desktop device sends an "account metadata" message that includes information about the account being connected to, such as:
+
+```
+the user's email address
+the user's profile picture and display name
+the device's display name.
+Additional server derived information potentially including:
+Counterpart's User Agent information
+Counterpart's GeoIP derived location
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: Removing this part, it's just repeating the messages section below

File: docs/pairing-architecture.md [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+ * The channel key is what secures the pairing flow.
+ * The channel id and channel key are encoded using base64url and then combined into a "pairing url". 
+ * The generated URL looks like: https://accounts.firefox.com/pair#channel_id=[channel_id]&channel_key=[channel_key]
+ * This URL is then encoded in a QR code and displayed on screen.
+ 
+#### 2) Mobile scans the QR code and connects to the encrypted channel
+
+To initiate pairing, the Mobile device scans the QR code displayed by Desktop and extracts the channel id and channel key.
+It opens a websocket to the Channel Server using the same fxa-pairing-channel library specifying the channel id, channel key.
+The Channel Server will now relay websocket messages from Mobile to Desktop and vice-versa. The supplicant broker handles the
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: 👍

File: docs/pairing-architecture.md [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+ * receiving the OAuth request from the Mobile device
+ * constructing the keys_jwe bundle and authorizing the OAuth request
+ * return the OAuth code to the Mobile device over the websocket channel
+ * communicating between Native code and Authority web content using WebChannels
+
+### QR Code and URLs
+
+The QR code is a URL: `https://accounts.firefox.com/pair/#channel_id=XY&channel_key=AB`
+
+Combined with the OAuth flow it results in the following:
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/pairing-architecture.md]: 👍

**Comment by vladikoff** [comment]
@rfk @shane-tomlinson big clean up complete!

**Pull request merged**